src/fetch-groups.js

```diff
@@ -0,0 +1,157 @@
+/**
+ * Module that exports a function that takes a list of specifications as input
+ * and computes, for each of them, the name of the organization and groups
+ * within that organization that develop the specification.
+ *
+ * The function needs an authentication token for the GitHub API as well as for
+ * the W3C API.
+ */
+
+const fetch = require("node-fetch");
+const { Octokit } = require("@octokit/rest");
+
+
+/**
+ * Exports main function that takes a list of specs (with a url property)
+ * as input, completes entries with an "organization" property that contains the
+ * name of the organization such as W3C, WHATWG, IETF, Khronos Group,
+ * Ecma International, and a "groups" property that contains an array of objects
+ * that describe the groups responsible for the spec.
+ * 
+ * The function preserves the properties if they have already been provided in
+ * the input array.
+ *
+ * The options parameter is used to specify the GitHub API and W3C API
+ * authentication tokens.
+ */
+module.exports = async function (specs, options) {
+  // Maintain a cache of fetched resources in memory to avoid sending the
+  // same fetch request again and again
+  const cache = {};
+
+  // Helper function to retrieve a JSON resource or return null if resource
+  // cannot be retrieved
+  async function fetchJSON(url, options) {
+    const body = cache[url] ?? await fetch(url, options).then(res => {
+      if (res.status !== 200) {
+        throw new Error(`W3C API returned an error, status code is ${res.status}`);
+      }
+      return res.json();
+    });
+    cache[url] = body;
+    return body;
+  }
+
+  const w3cOptions = options?.w3cApiKey ? { headers: {
+    Authorization: `W3C-API apikey="${options.w3cApiKey}"`
+  }} : {};
+
+  for (const spec of specs) {
+    // Note: same RegExp logic to detect WHATWG, TC39, SVG, Khronos Group and
+    // GitHub URLs as in compute-repository.
```

Comment:
is there an opportunity to factor that logic out to reduce the maintenance burden?

Reply:
Done in latest commit.